[sonic-package-manager] first phase implementation of sonic-package-manager (#1527)
Azure/sonic-utilities: doc/Command-Reference.md

## Software Installation and Management

> SONiC software image can be installed in two methods, viz, "using sonic-installer tool", "ONIE Installer".

Reviewer: Suggest removing the word "software", making a list, and also mentioning other bootloaders. E.g.,

```suggestion
SONiC image can be installed in one of two methods:
1. From within a running SONiC iamge using the `sonic-installer` utility
2. From the vendor's bootloader (E.g., ONIE,  Aboot, etc.)
```

SONiC feature Docker images (aka "SONiC packages") available to be installed with *sonic-package-manager* utility.

### SONiC Package Manager

This is a command line tool that provides functionality to manage SONiC Packages on SONiC device.

**sonic-package-manager list**

This command lists all available SONiC packages, their desription, installed version and installation status.
SONiC package status can be *Installed*, *Not installed* or *Built-In*. "Built-In" status means that a feature is built-in to SONiC image and can't be upgraded or uninstalled.

- Usage:
  ```

>   sonic-package-manager list

Reviewer 2: Guess "sonic"  is implied. It could be part of show command

"package list"

Author: This is on par with "sonic-installer".
It can be later added in show command, similar to "show boot", which calls "sonic-installer list".

  macsec          docker-macsec                SONiC macsec package          1.0.0      Built-In
  mgmt-framework  docker-sonic-mgmt-framework  SONiC mgmt-framework package  1.0.0      Built-In
  nat             docker-nat                   SONiC nat package             1.0.0      Built-In
  pmon            docker-platform-monitor      SONiC pmon package            1.0.0      Built-In
  radv            docker-router-advertiser     SONiC radv package            1.0.0      Built-In
  sflow           docker-sflow                 SONiC sflow package           1.0.0      Built-In
  snmp            docker-snmp                  SONiC snmp package            1.0.0      Built-In
  swss            docker-orchagent             SONiC swss package            1.0.0      Built-In
  syncd           docker-syncd-mlnx            SONiC syncd package           1.0.0      Built-In
  teamd           docker-teamd                 SONiC teamd package           1.0.0      Built-In
  telemetry       docker-sonic-telemetry       SONiC telemetry package       1.0.0      Built-In
  ```

**sonic-package-manager repository add**

> 

Reviewer 2: Could it be simply, "package add ..."

Author: This is as in HLD.
A package is a contrete Docker image, while "repository add" adds a repository where many packages can be stored, so I don't think "package add" is a right command name.

This command will add a new entry in the package database. The package has to be *Not Installed* in order to be removed from package database. *NOTE*: this command requires elevated (root) privileges to run.

- Usage:
  ```
  Usage: sonic-package-manager repository add [OPTIONS] NAME REPOSITORY

>     Add a new repository to database.

Reviewer 2: Not adding a repo, but a new III party package. Is that correct?

Author: Adding a repo is correct. A package is a docker image, a repo may contain several docker images.

  Options:
    --default-reference TEXT  Default installation reference.
    --description TEXT        Optional package entry description.
    --help                    Show this message and exit.
  ```
- Example:
  ```
  admin@sonic:~$ sudo sonic-package-manager repository add \
    cpu-report azure/sonic-cpu-report --default-reference 1.0.0
  ```

**sonic-package-manager repository remove**

> This command will remove an entry from the package database. *NOTE*: this command requires elevated (root) privileges to run.

Reviewer 2: What would be the case, if package is running ?

Author: This is disallowed by package-manager code. Please check https://github.com/Azure/sonic-utilities/pull/1527/files#diff-d1e2ce441973ddcb53f8fe1cd3d2e3dcae8aa8ca30fca3fe2336233b7e488d5eR128-R134

- Usage:
  ```
  Usage: sonic-package-manager repository remove [OPTIONS] NAME

    Remove package from database.

  Options:
    --help  Show this message and exit.
  ```
- Example:
  ```
  admin@sonic:~$ sudo sonic-package-manager repository remove cpu-report
  ```

**sonic-package-manager install**

This command pulls and installs package on SONiC host. *NOTE*: this command requires elevated (root) privileges to run.

> 

Reviewer 2: Make it explicit "this is only for default/added packages"

would you require "-f", when  you are downgrading ?

What would be the user faced behavior, if the package is currently running ?

Is there any option for user to revert back to what he had before ?

Author: "this is only for default/added packages" - what does it mean? You mean that the user first needs to add a repository?

"would you require "-f", when you are downgrading ?" - yes, please see the comment block in the code with some thoughts on  downgrade - https://github.com/Azure/sonic-utilities/pull/1527/files#diff-a8e54e07d62d4629cbfd9ca9c289208aa437f0d8f722a8f8b363df9c38b315e6R475-R488

"What would be the user faced behavior, if the package is currently running ?". If a package is currently running, meaning it is installed, installing the same package again (or different version of the same package) will be blocked https://github.com/Azure/sonic-utilities/pull/1527/files#diff-a8e54e07d62d4629cbfd9ca9c289208aa437f0d8f722a8f8b363df9c38b315e6R322-R325

"Is there any option for user to revert back to what he had before ?" - no, user is required to downgrade with force.

Author: BTW, the default versions that came with the image are recorded in "default_reference" in packages.json, so user can always look for default and reinstall the one that was planned to be used for his sonic image

Author: We can add a command to reset the package to a default version:
```
reset-default x == upgrade x==<default_reference>
```

                                    NOTE: This argument is mutually exclusive
                                    with arguments: [from_repository,
                                    package_expr].

    -f, --force                     Force operation by ignoring failures
    -y, --yes                       Automatically answer yes on prompts
    -v, --verbosity LVL             Either CRITICAL, ERROR, WARNING, INFO or
                                    DEBUG

    --skip-cli-plugin-installation  Do not install CLI plugins provided by the
                                    package on the host OS. NOTE: In case when
                                    package /cli/mandatory field is set to True
                                    this option will fail the installation.

    --help                          Show this message and exit.
  ```
- Example:
  ```
  admin@sonic:~$ sudo sonic-package-manager install dhcp-relay==1.0.2
  ```
  ```
  admin@sonic:~$ sudo sonic-package-manager install --from-repository azure/sonic-cpu-report:latest
  ```
  ```
  admin@sonic:~$ sudo sonic-package-manager install --from-tarball sonic-docker-image.gz
  ```

**sonic-package-manager uninstall**

> This command uninstalls package from SONiC host. *NOTE*: this command requires elevated (root) privileges to run.

Reviewer 2: Does this bring back, original copy for images that are part of install ?

Author: No, that would mean the switch has to save previous docker images and may take more disk space.
User can uninstall and install again older version if he would like to rollback.

This was not previously discussed as a requirement, so may I suggest not to add more functionality as part of this PR?
Besides, user is not blocked to uninstall and install or downgrade to previous versions of packages.

This PR is phase 1 and trying to provide minimum to get the 3rd party docker up and runnning on SONiC, while the rollback (previous comment) and original images retention seems like somthing we can build on top of the minimum which we defined for phase 1.

- Usage:
  ```
  Usage: sonic-package-manager uninstall [OPTIONS] NAME

    Uninstall package

  Options:
    -f, --force          Force operation by ignoring failures
    -y, --yes            Automatically answer yes on prompts
    -v, --verbosity LVL  Either CRITICAL, ERROR, WARNING, INFO or DEBUG
    --help               Show this message and exit.

  ```
- Example:
  ```
  admin@sonic:~$ sudo sonic-package-manager uninstall dhcp-relay
  ```

**sonic-package-manager upgrade**

> This command upgrades package on SONiC host to a newer version. The procedure of upgrading a package will restart the corresponding service. *NOTE*: this command requires elevated (root) privileges to run.

Reviewer 2: What is the diff between this and default install, that brings in latest?

Author: "install" command does not bring in the latest by default. "install" will look for "default-reference" for this package in "packages.json" and install from that tag.

For the difference between install and upgrade please refer to HLD - https://github.com/stepanblyschak/SONiC/blob/sonic-app-ext-3/doc/sonic-application-extention/sonic-application-extention-hld.md#sonic-package-upgrade-flow.

Reviewer 3: Following "apt" again, i think "upgrade" should not take arguments, but instead upgrade all packages to their latest available versions (which means we maybe also need an "sonic-package-manager update" to update all repository sources.

Author: "sonic-package-manager update" - this requires some external source of packages.json that will be updated - this is not planned yet.

Upgrade to their latest available versions - this might not be safe, because "latest" may be incompatible. Currently user is required to explicitely specify the version he would like to have.

                                    with arguments: [package_expr,
                                    from_repository].

    -f, --force                     Force operation by ignoring failures
    -y, --yes                       Automatically answer yes on prompts
    -v, --verbosity LVL             Either CRITICAL, ERROR, WARNING, INFO or
                                    DEBUG

    --skip-cli-plugin-installation  Do not install CLI plugins provided by the
                                    package on the host OS. NOTE: In case when
                                    package /cli/mandatory field is set to True
                                    this option will fail the installation.

    --help                          Show this message and exit.
  ```
- Example:
  ```
  admin@sonic:~$ sudo sonic-package-manager upgrade dhcp-relay==2.0.0
  ```
  ```
  admin@sonic:~$ sudo sonic-package-manager upgrade --from-repository azure/sonic-cpu-report:latest
  ```
  ```
  admin@sonic:~$ sudo sonic-package-manager upgrade --from-tarball sonic-docker-image.gz
  ```

**sonic-package-manager show package versions**

> 

Reviewer 2: Is this the list of all images locally available in the device ?

If yes, does this mean, each time a image is upgraded, the older one is retained ?
else which repo being referred?

    --plain  Plain output
    --help   Show this message and exit.
  ```
- Example:
  ```
  admin@sonic:~$ sonic-package-manager show package versions dhcp-relay
  • 1.0.0
  • 1.0.2
  • 2.0.0
  ```

**sonic-package-manager show package changelog**

This command fetches the changelog from package manifest and displays it. *NOTE*: package changelog can be retrieved from registry or read from image tarball without installing it.

- Usage:
  ```
  Usage: sonic-package-manager show package changelog [OPTIONS] [PACKAGE_EXPR]

    Print package changelog

  Options:
    --from-repository TEXT  Fetch package directly from image registry
                            repository NOTE: This argument is mutually exclusive
                            with arguments: [from_tarball, package_expr].
    --from-tarball FILE     Fetch package from saved image tarball NOTE: This
                            argument is mutually exclusive with arguments:
                            [package_expr, from_repository].
    --help                  Show this message and exit.
  ```
- Example:
  ```
  admin@sonic:~$ sonic-package-manager show package changelog dhcp-relay

>   1.0.0:

Reviewer 2: Where this data from ?

Author: Changelog is from manifest. Please refer to HLD https://github.com/stepanblyschak/SONiC/blob/sonic-app-ext-3/doc/sonic-application-extention/sonic-application-extention-hld.md#sonic-package-changelog

Reviewer 2: manifest.json.j2 does not accommodate yet

    • Initial release

        Author (author@email.com) Mon, 25 May 2020 12:25:00 +0300
  ```

**sonic-package-manager show package manifest**

This command fetches the package manifest and displays it. *NOTE*: package manifest can be retrieved from registry or read from image tarball without installing it.

- Usage:

>   ```

Reviewer 2: How does one specify repo to use ?

Author: "show package manifest some-package==1.0.2" - this will access the repository recorded for package "sonic-package" in packages.json

Another option:
"show package manifest --from-repository <some-repo>:<some-tag>"

## Software Installation and Management

SONiC image can be installed in one of two methods:

> 1. From within a running SONiC iamge using the `sonic-installer` utility

Author: Done

  lldp            docker-lldp                  SONiC lldp package            1.0.0      Built-In
  macsec          docker-macsec                SONiC macsec package          1.0.0      Built-In
  mgmt-framework  docker-sonic-mgmt-framework  SONiC mgmt-framework package  1.0.0      Built-In
  nat             docker-nat                   SONiC nat package             1.0.0      Built-In
  pmon            docker-platform-monitor      SONiC pmon package            1.0.0      Built-In
  radv            docker-router-advertiser     SONiC radv package            1.0.0      Built-In
  sflow           docker-sflow                 SONiC sflow package           1.0.0      Built-In
  snmp            docker-snmp                  SONiC snmp package            1.0.0      Built-In
  swss            docker-orchagent             SONiC swss package            1.0.0      Built-In
  syncd           docker-syncd-mlnx            SONiC syncd package           1.0.0      Built-In
  teamd           docker-teamd                 SONiC teamd package           1.0.0      Built-In
  telemetry       docker-sonic-telemetry       SONiC telemetry package       1.0.0      Built-In

>   ```

Author: We may add if this is required

>   ```

Author: It can be added

**sonic-package-manager repository add**

> This command will add a new entry in the package database. The package has to be *Not Installed* in order to be removed from package database. *NOTE*: this command requires elevated (root) privileges to run.

Author: This is a comment to outdated file.

- Usage:
  ```
  Usage: sonic-package-manager repository add [OPTIONS] NAME REPOSITORY

    Add a new repository to database.

  Options:
    --default-reference TEXT  Default installation reference.
    --description TEXT        Optional package entry description.
    --help                    Show this message and exit.
  ```
- Example:
  ```
  admin@sonic:~$ sudo sonic-package-manager repository add \

>     cpu-report azure/sonic-cpu-report --default-reference 1.0.0

Author: what is this "azure/sonic-cpu-report" ?

This is a repository, arbitrary example.

What is the purpose of default-reference ?

The purpose of default-reference is described in HLD - https://github.com/stepanblyschak/SONiC/blob/sonic-app-ext-3/doc/sonic-application-extention/sonic-application-extention-hld.md#sonic-package-database.

The install command, takes even repo path, so wonder what is the value of this repo?

Not sure I understand the question.
The install command takes an expression in format name==version. The reposiotory lookup takes place based on name of the package.

  ```

**sonic-package-manager repository remove**

This command will remove an entry from the package database. *NOTE*: this command requires elevated (root) privileges to run.

- Usage:
  ```
  Usage: sonic-package-manager repository remove [OPTIONS] NAME

    Remove package from database.

  Options:
    --help  Show this message and exit.
  ```
- Example:
  ```
  admin@sonic:~$ sudo sonic-package-manager repository remove cpu-report
  ```

**sonic-package-manager install**

This command pulls and installs package on SONiC host. *NOTE*: this command requires elevated (root) privileges to run.

- Usage:
  ```
  Usage: sonic-package-manager install [OPTIONS] [PACKAGE_EXPR]

>     Install package

Author: Updated the help for this command.

The default install (w/o version) could be potentially downgrading. Will there be a WARNING?
This will have to be made explicit.
[SB] Downgrade by default is rejected unless --allow-downgrade is provided.

For a smooth user experience, the default should install "latest" -- My view
[SB] It cannot be smooth IMO in that case. User running a specific image wants by default a version that was validated and tested against that image instead of taking latest version. If user wants, user is free to specify latest.

  Options:
    --enable                        Set the default state of the feature to
                                    enabled and enable feature right after
                                    installation. NOTE: user needs to execute
                                    "config save -y" to make this setting
                                    persistent

>     --default-owner [local|kube]    Default owner configuration setting for a

Author: This is changed to set-owner

                                    feature

>     --from-repository TEXT          Fetch package directly from image registry

Author: Either [PACKAGE_EXPR] or --from-repository or --from-tarball is required

                                    repository NOTE: This argument is mutually
                                    exclusive with arguments: [from_tarball,
                                    package_expr].

    --from-tarball FILE             Fetch package from saved image tarball
                                    NOTE: This argument is mutually exclusive
                                    with arguments: [from_repository,
                                    package_expr].

>     -f, --force                     Force operation by ignoring failures

Author: Added to this flag description

    -y, --yes                       Automatically answer yes on prompts

>     -v, --verbosity LVL             Either CRITICAL, ERROR, WARNING, INFO or

Author: Added default

    -v, --verbosity LVL             Either CRITICAL, ERROR, WARNING, INFO or
                                    DEBUG

    --skip-cli-plugin-installation  Do not install CLI plugins provided by the
                                    package on the host OS. NOTE: In case when
                                    package /cli/mandatory field is set to True
                                    this option will fail the installation.

    --help                          Show this message and exit.
  ```
- Example:
  ```
  admin@sonic:~$ sudo sonic-package-manager upgrade dhcp-relay==2.0.0
  ```
  ```
  admin@sonic:~$ sudo sonic-package-manager upgrade --from-repository azure/sonic-cpu-report:latest
  ```
  ```
  admin@sonic:~$ sudo sonic-package-manager upgrade --from-tarball sonic-docker-image.gz
  ```

**sonic-package-manager show package versions**

This command will access repository for corresponding package and retrieve a list of available versions.

- Usage:
  ```
  Usage: sonic-package-manager show package versions [OPTIONS] NAME

    Print available versions

  Options:
    --all    Show all available tags in repository

>     --plain  Plain output

Author: Does this also stop and uninstall the corresponding service, if running? [SB] It is user responsibilty

Or would it demand, the service should be in stopped state, for remove to be success? [SB] Yes

Would it fail, if there are dependent packages installed? [SB] Yes, uninstallation will fail.

>     --plain  Plain output

Author: Either we register the repo, in which case don't allow install/upgrade to override
Or, allow this command to accept repo

Could you please clarify which command you are refering to and what is your suggestion?

    --help   Show this message and exit.
  ```
- Example:
  ```
  admin@sonic:~$ sonic-package-manager show package versions dhcp-relay
  • 1.0.0
  • 1.0.2
  • 2.0.0
  ```

**sonic-package-manager show package changelog**

This command fetches the changelog from package manifest and displays it. *NOTE*: package changelog can be retrieved from registry or read from image tarball without installing it.

- Usage:
  ```
  Usage: sonic-package-manager show package changelog [OPTIONS] [PACKAGE_EXPR]

    Print package changelog

  Options:
    --from-repository TEXT  Fetch package directly from image registry
                            repository NOTE: This argument is mutually exclusive

>                             with arguments: [from_tarball, package_expr].

Author: Does manifest carry change log ? [SB] Yes
manifest.json.j2 does not have it. [SB] We may extend, btw, extension usually won't be in sonic-buildimage repository so they won't use manifest.json.j2

    Add a new repository to database. Repository in Docker Registry V2.

  Options:
    --default-reference TEXT  Default installation reference
    --description TEXT        Optional package entry description
    --help                    Show this message and exit.
  ```
- Example:
  ```
  admin@sonic:~$ sudo sonic-package-manager repository add \
    cpu-report azure/sonic-cpu-report --default-reference 1.0.0
  ```

**sonic-package-manager repository remove**

This command will remove an entry from the package database. The package has to be *Not Installed* in order to be removed from package database. *NOTE*: this command requires elevated (root) privileges to run.

- Usage:
  ```
  Usage: sonic-package-manager repository remove [OPTIONS] NAME

    Remove package from database.

  Options:
    --help  Show this message and exit.
  ```
- Example:
  ```
  admin@sonic:~$ sudo sonic-package-manager repository remove cpu-report
  ```

**sonic-package-manager install**

This command pulls and installs package on SONiC host. *NOTE*: this command requires elevated (root) privileges to run.

- Usage:
  ```
  Usage: sonic-package-manager install [OPTIONS] [PACKAGE_EXPR]

>     Install package using [PACKAGE_EXPR] in format "<name>==<version>" or "<name>@<tag|digest>"

Reviewer 3: ```suggestion
    Install package
    [PACKAGE_EXPR] can be in the format "<name>=<version>" or "<name>@<tag|digest>"
```

Reviewer 3: When looking at other package managers, most allow installing multiple packages at the same time by chaining the arguments. Would it be possible to also allow this here?

Author: It can be added

  Options:

>     --enable                      Set the default state of the feature to

Author: Controls the feature state - https://github.com/Azure/SONiC/blob/master/doc/Optional-Feature-Control.md#sonic-optional-feature-control-enhancement

                                  enabled and enable feature right after
                                  installation. NOTE: user needs to execute
                                  "config save -y" to make this setting
                                  persistent
    --default-owner [local|kube]  Default owner configuration setting for a
                                  feature
    --from-repository TEXT        Fetch package directly from image registry
                                  repository NOTE: This argument is mutually
                                  exclusive with arguments: [package_expr,
                                  from_tarball].

>     --from-tarball FILE           Fetch package from saved image tarball NOTE:

Reviewer 3: If i am allowed to install "from tarball", why am i not allowed to remove a repository before removing (L8030) ?

Author: Installing from tarball will add an entry in packages.json. So still your not able to remove this record.

                                  This argument is mutually exclusive with
                                  arguments: [package_expr, from_repository].

>     -f, --force                   Force operation by ignoring failures

Reviewer 3: I think a "force install" is very hard to implement, given that we do not know why the "not-forced" approach did not work. Maybe we should drop this flag and just let the user know why this did not work and how she/he can fix that.

Author: "force install" is very hard to implement - could you please give an example? Could you please take a look at its implementation?

We cannot drop it, it was a requirement.

    -y, --yes                     Automatically answer yes on prompts

>     -v, --verbosity LVL           Either CRITICAL, ERROR, WARNING, INFO or DEBUG

Reviewer 3: This should mention what the default level is.

Author: This help comes from https://pypi.org/project/click-log/ click decorator, I cannot modify it unless I copy and modify its code.

    --skip-host-plugins           Do not install host OS plugins provided by the
                                  package (CLI, etc). NOTE: In case when package
                                  host OS plugins are set as mandatory in
                                  package manifest this option will fail the
                                  installation.
    --help                        Show this message and exit.
  ```
- Example:
  ```

>   admin@sonic:~$ sudo sonic-package-manager install dhcp-relay==1.0.2

Reviewer 3: This would need to be aligned if we decide to use only one "="

  ```
  ```
  admin@sonic:~$ sudo sonic-package-manager install --from-repository azure/sonic-cpu-report:latest
  ```
  ```
  admin@sonic:~$ sudo sonic-package-manager install --from-tarball sonic-docker-image.gz
  ```

**sonic-package-manager uninstall**

This command uninstalls package from SONiC host. *NOTE*: this command requires elevated (root) privileges to run.

- Usage:
  ```
  Usage: sonic-package-manager uninstall [OPTIONS] NAME

>     Uninstall package

Reviewer 3: ```suggestion
    Uninstall package
```

Reviewer 3: Should we also allow uninstalling multiple packages at once?

Author: It can be added

  Options:

>     -f, --force          Force operation by ignoring failures

Reviewer 3: Same as for install, i think force is not a good idea here. The package-manager should abstract and "protect the user" in the sense, that it does not break the system. If the user decides to replace a docker container or binary somewhere, she/he is still free to do so without the package-manger.

Author: The package-manager should abstract and "protect the user" in the sense, that it does not break the system. 
[SB] Then the user does not use --force

    -y, --yes            Automatically answer yes on prompts

>     -v, --verbosity LVL  Either CRITICAL, ERROR, WARNING, INFO or DEBUG

Reviewer 3: should state the default level

                            with arguments: [package_expr, from_tarball].
    --from-tarball FILE     Fetch package from saved image tarball NOTE: This
                            argument is mutually exclusive with arguments:
                            [from_repository, package_expr].
    -f, --force             Force operation by ignoring failures
    -y, --yes               Automatically answer yes on prompts
    -v, --verbosity LVL     Either CRITICAL, ERROR, WARNING, INFO or DEBUG
    --skip-host-plugins     Do not install host OS plugins provided by the
                            package (CLI, etc). NOTE: In case when package host
                            OS plugins are set as mandatory in package manifest
                            this option will fail the installation.
    --help                  Show this message and exit.
  ```
- Example:
  ```
  admin@sonic:~$ sudo sonic-package-manager upgrade dhcp-relay==2.0.0
  ```
  ```
  admin@sonic:~$ sudo sonic-package-manager upgrade --from-repository azure/sonic-cpu-report:latest
  ```
  ```
  admin@sonic:~$ sudo sonic-package-manager upgrade --from-tarball sonic-docker-image.gz
  ```

> **sonic-package-manager reset**

Reviewer 3: Personally, i find "reset" very confusing when targeting "just one package". I think a command to reset ALL packages to their default state (if possible when considering that those packages might also have installed tools on the host, which usually do not have a straight "revert" option) would be very useful thou.

Author: "packages might also have installed tools on the host, which usually do not have a straight "revert" option" - could you please give an example?
Reset multiple can be added

> This comamnd resets the package by reinstalling it to its default version.

Reviewer 3: ```suggestion
This command resets the package by reinstalling it to its default version.
```

- Usage:
  ```
  Usage: sonic-package-manager reset [OPTIONS] NAME

    Reset package to the default version

  Options:
    -f, --force          Force operation by ignoring failures
    -y, --yes            Automatically answer yes on prompts
    -v, --verbosity LVL  Either CRITICAL, ERROR, WARNING, INFO or DEBUG
    --skip-host-plugins  Do not install host OS plugins provided by the package
                        (CLI, etc). NOTE: In case when package host OS plugins
                        are set as mandatory in package manifest this option
                        will fail the installation.
    --help               Show this message and exit.
  ```
- Example:
  ```
  admin@sonic:~$ sudo sonic-package-manager reset dhcp-relay
  ```

**sonic-package-manager show package versions**

> This command will access repository for corresponding package and retrieve a list of available versions.

Reviewer 3: ```suggestion
This command will retrieve a list of all available versions for the given package from the configured upstream repository
```

- Usage:
  ```
  Usage: sonic-package-manager show package versions [OPTIONS] NAME

>     Print available versions

Reviewer 3: ```suggestion
    Show available versions
```

  Options:

>     --all    Show all available tags in repository

Reviewer 3: I think "tags" might confuse people here, since it was not mentioned in the docs so far.

Author: The docs states about Docker and Repository and Registry, it does not look like user can be aware about it but not understand "tags in repository"

>     --all    Show all available tags in repository

Reviewer 3: ```suggestion
    --all    Show all versions available in the repository
```

Reviewer 3: Is "--all" the default ? If not, what is the default ?

Author: No, it is not. --all is an option so it is not a default.

>     --plain  Plain output

Reviewer 3: what does "--plain" mean here? which format is that?

Author: Added example

    --help   Show this message and exit.
  ```
- Example:
  ```

>   admin@sonic:~$ sonic-package-manager show package versions dhcp-relay

Reviewer 3: Although it looks nice to humans, output like this is very hard to parse for automation tools. Maybe we can offer an option to output something like yaml, json or something similar here?

Author: Then use --plain

Reviewer 3: Which format will "plain" be in ? If it is any defined format (like yaml, json or something alike), i think we wanna rename that option to "--format $FORMAT" or "--$FORMAT"

  • 1.0.0
  • 1.0.2
  • 2.0.0
  ```

**sonic-package-manager show package changelog**

> This command fetches the changelog from package manifest and displays it. *NOTE*: package changelog can be retrieved from registry or read from image tarball without installing it.

Reviewer 3: ```suggestion
This command fetches the changelog from the package manifest and displays it. *NOTE*: package changelog can be retrieved from registry or read from image tarball without installing it.
```

- Usage:
  ```
  Usage: sonic-package-manager show package changelog [OPTIONS] [PACKAGE_EXPR]

>     Print package changelog

Reviewer 3: ```suggestion
    Show package changelog
```

  Options:
    --from-repository TEXT  Fetch package directly from image registry
                            repository NOTE: This argument is mutually exclusive
                            with arguments: [from_tarball, package_expr].
    --from-tarball FILE     Fetch package from saved image tarball NOTE: This
                            argument is mutually exclusive with arguments:
                            [package_expr, from_repository].
    --help                  Show this message and exit.
  ```
- Example:
  ```
  admin@sonic:~$ sonic-package-manager show package changelog dhcp-relay
  1.0.0:

    • Initial release

        Author (author@email.com) Mon, 25 May 2020 12:25:00 +0300
  ```

**sonic-package-manager show package manifest**

This command fetches the package manifest and displays it. *NOTE*: package manifest can be retrieved from registry or read from image tarball without installing it.

- Usage:
  ```
  Usage: sonic-package-manager show package manifest [OPTIONS] [PACKAGE_EXPR]

>     Print package manifest content

Reviewer 3: ```suggestion
    Show package manifest
```

  Done
  ```

SONiC image installation will install SONiC packages that are installed in currently running SONiC image. In order to perform clean SONiC installation use *--skip-package-migration* option when installing SONiC image:

- Example:
  ```

>   admin@sonic:~$ sudo sonic-installer install https://sonic-jenkins.westus.cloudapp.azure.com/job/xxxx/job/buildimage-xxxx-all/xxx/artifact/target/sonic-xxxx.bin --skip-package-migration

Reviewer 2: Will it upgrade packages, if currently installed is lower version, w/o --skip flag?

Author: yes, please refer to HLD - https://github.com/stepanblyschak/SONiC/blob/sonic-app-ext-3/doc/sonic-application-extention/sonic-application-extention-hld.md#sonic-2-sonic-upgrade

  Done
  ```

> SONiC image installation will install SONiC packages that are installed in currently running SONiC image. In order to perform clean SONiC installation use *--skip-package-migration* option when installing SONiC image:

Reviewer 3: ```suggestion
Installing a new image using the sonic-installer will keep using the packages installed on the currently running SONiC image and automatically migrate those. In order to perform clean SONiC installation use the *--skip-package-migration* option:
```
